flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DeclarativeSlotManager.java

```diff
@@ -0,0 +1,609 @@
+			slotId,
+			jobId,
+			new AllocationID(),
+			resourceProfile,
+			targetAddress,
+			resourceManagerId,
+			taskManagerRequestTimeout);
+
+		requestFuture.whenComplete(
+			(Acknowledge acknowledge, Throwable throwable) -> {
+				if (acknowledge != null) {
+					completableFuture.complete(acknowledge);
+				} else {
+					completableFuture.completeExceptionally(throwable);
+				}
+			});
+
+		completableFuture.whenCompleteAsync(
+			(Acknowledge acknowledge, Throwable throwable) -> {
+				try {
+					if (acknowledge != null) {
+						LOG.trace("Completed allocation of slot {} for job {}.", slotId, jobId);
+						slotTracker.notifyAllocationComplete(slotId, jobId);
+					} else {
+						if (throwable instanceof SlotOccupiedException) {
+							SlotOccupiedException exception = (SlotOccupiedException) throwable;
+							LOG.debug("Tried allocating slot {} for job {}, but it was already allocated for job {}.", slotId, jobId, exception.getJobId());
+							// report as a slot status to force the state transition
+							// this could be a problem if we ever assume that the task executor always reports about all slots
+							slotTracker.notifySlotStatus(Collections.singleton(new SlotStatus(slotId, taskManagerSlot.getResourceProfile(), exception.getJobId(), exception.getAllocationId())));
+						} else {
+							if (throwable instanceof CancellationException) {
+								LOG.debug("Cancelled allocation of slot {} for job {}.", slotId, jobId, throwable);
+							} else {
+								LOG.warn("Slot allocation for slot {} for job {} failed.", slotId, jobId, throwable);
+								slotTracker.notifyFree(slotId);
+							}
+						}
+					}
+					checkResourceRequirements();
+				} catch (Exception e) {
+					LOG.error("Error while completing the slot allocation.", e);
+				}
+			},
+			mainThreadExecutor);
+	}
+
+	private void tryFulfillRequirementsWithPendingOrNewSlots(JobID jobId, Map<ResourceProfile, Integer> missingResources, ResourceCounter pendingSlots) {
+		for (Map.Entry<ResourceProfile, Integer> missingResource : missingResources.entrySet()) {
+			ResourceProfile profile = missingResource.getKey();
+			for (int i = 0; i < missingResource.getValue(); i++) {
+				if (!tryFulfillWithPendingSlots(profile, pendingSlots)) {
+					Optional<ResourceRequirement> newlyFulfillableRequirements = taskExecutorManager.allocateWorker(profile);
+					if (newlyFulfillableRequirements.isPresent()) {
+						ResourceRequirement newSlots = newlyFulfillableRequirements.get();
+						// reserve one of the new slots
+						if (newSlots.getNumberOfRequiredSlots() > 1) {
+							pendingSlots.incrementCount(newSlots.getResourceProfile(), newSlots.getNumberOfRequiredSlots() - 1);
+						}
+					} else {
+						LOG.warn("Could not fulfill resource requirements of job {}.", jobId);
+						resourceActions.notifyNotEnoughResourcesAvailable(jobId, resourceTracker.getAcquiredResources(jobId));
+						return;
+					}
```

Comment:
One could factor this part out into a separate method which might make it a bit less complex.